File: cpp/binary_search_tree/splay-tree.cpp
struct Node {
  Node *left, *right;
  ll key;
  int count;

  Node(ll key, int count)
    : key(key), count(count), left(nullptr), right(nullptr) {}
};

// splay a node nd
Node* splay(stack<Node*> &st, stack<bool> &dr, Node* nd) {
  Node *l = nd->left, *r = nd->right;
  int L = (l ? l->count : 0), R = (r ? r->count : 0);
  int c = st.size() >> 1;
  Node *x, *y; bool d, d1;
  while(c--) {
    x = st.top(); st.pop();
    y = st.top(); st.pop();
    d = dr.top(); dr.pop();
    d1 = dr.top(); dr.pop();

    if(d == d1) {
      // Zig-zig step
      int e = (y->count -= L + R + 2);
      if(d) {
        y->right = x->left; x->left = y;
        x->right = l; l = x;

        l->count = (L += e + 1);
      } else {
        y->left = x->right; x->right = y;
        x->left = r; r = x;

        r->count = (R += e + 1);
      }
    } else {
      // Zig-zag step
      if(d) {
        x->right = l; l = x;
        y->left = r; r = y;

        L = (l->count -= R + 1);
        R = (r->count -= L + 1);
      } else {
        x->left = r; r = x;
        y->right = l; l = y;

        R = (r->count -= L + 1);
        L = (l->count -= R + 1);
      }
    }
  }

  if(!st.empty()) {
    // Zig step
    Node *x = st.top(); bool d = dr.top();
    if(d) {
      x->right = l; l = x;
      L = (l->count -= R + 1);
    } else {
      x->left = r; r = x;
      R = (r->count -= L + 1);
    }
  }

  nd->left = l; nd->right = r;
  nd->count = L + R + 1;

  return nd;
}

// merge a tree l with a tree r
Node* t_merge(Node *l, Node *r) {
  if(!l || !r) {
    return l ? l : r;
  }

  if(!l->right) {
    l->count += r->count;
    l->right = r;
    return l;
  }

  stack<Node*> st;
  stack<bool> dr;
  Node *x = l;
  while(x->right) {
    st.push(x); dr.push(true);
    x = x->right;
  }

  l = splay(st, dr, x);
  l->count += r->count;
  l->right = r;
  return l;
}

// split a tree t into two trees of size k and |t|-k
typedef pair<Node*, Node*> NodeP;
NodeP t_split(Node *t, int k) {
  if(!t || !(0 < k && k < t->count)) {
    return NodeP(t, nullptr);
  }
  Node *x = t;
  stack<Node*> st;
  stack<bool> dr;
  while(x) {
    Node *l = x->left;
    int c = (l ? l->count : 0) + 1;

    if(c == k) break;

    st.push(x);
    if(c < k) {
      k -= c;
      x = x->right;
      dr.push(true);
    } else {
      x = x->left;
      dr.push(false);
    }
  }
  Node *l = splay(st, dr, x), *r = l->right;
  if(r) l->count -= r->count;
  l->right = nullptr;
  return NodeP(l, r);
}

// insert a node with key = val
Node* insert(Node* t, ll val) {
  Node *nd = new Node(val, 1);
  stack<Node*> st;
  stack<bool> dr;
  Node *x = t, *y = nullptr;
  bool d;
  while(x) {
    st.push(y = x);
    dr.push(d = (x->key <= val));
    x->count++;
    x = (!d ? x->left : x->right);
  }
  if(!y) {
    return nd;
  }
  (!d ? y->left : y->right) = nd;
  return splay(st, dr, nd);
}

// delete a node with key = val
Node* remove(Node *t, ll val) {
  stack<Node*> st;
  stack<bool> dr;
  Node *x = t; bool d;
  while(x && x->key != val) {
    st.push(x);
    dr.push(d = (x->key <= val));
    x = (!d ? x->left : x->right);
  }
  if(!x) return t;

  t = splay(st, dr, x);
  return t_merge(t->left, t->right);
}

// find a node with key = val
Node* find(Node *t, ll val) {
  stack<Node*> st;
  stack<bool> dr;
  Node *x = t; bool d;
  while(x && x->key != val) {
    st.push(x);
    dr.push(d = (x->key <= val));
    x = (!d ? x->left : x->right);
  }
  if(!x) return t;

  return splay(st, dr, x);
}

// for debug
int dfs(Node *v, int k) {
  if(!v) return 0;
  int r = 0;
  if(v->left) r += dfs(v->left, k+1);
  rep(i, k) cout << " "; cout << v->key << " " << v->count << endl;
  if(v->right) r += dfs(v->right, k+1);
  return r+1;
}

File: python/binary_search_tree/splay-tree.py
# splay a node nd
def __splay(st, dr, nd):
    l = nd[0]; r = nd[1]
    L = (l[3] if l else 0); R = (r[3] if r else 0)
    c = len(st) >> 1
    while c:
        # y(d1)-x(d)-nd
        x = st.pop(); y = st.pop()
        d = dr.pop(); d1 = dr.pop()

        if d == d1:
            # Zig-zig step
            y[3] = e = y[3] - L - R - 2
            if d:
                y[1] = x[0]; x[0] = y
                x[1] = l; l = x

                l[3] = L = e + L + 1
            else:
                y[0] = x[1]; x[1] = y
                x[0] = r; r = x

                r[3] = R = e + R + 1
        else:
            # Zig-zag step
            if d:
                x[1] = l; l = x
                y[0] = r; r = y

                l[3] = L = l[3] - R - 1
                r[3] = R = r[3] - L - 1
            else:
                x[0] = r; r = x
                y[1] = l; l = y

                r[3] = R = r[3] - L - 1
                l[3] = L = l[3] - R - 1
        c -= 1
        # update(y); update(x)
    if st:
        # Zig step
        x = st[0]; d = dr[0]
        if d:
            x[1] = l; l = x
            l[3] = L = l[3] - R - 1
        else:
            x[0] = r; r = x
            r[3] = R = r[3] - L - 1
        # update(x)
    nd[0] = l; nd[1] = r
    nd[3] = L + R + 1
    # update(nd)
    return nd

# create a new node with key = val
def new_node(val):
    return [None, None, val, 1]

# insert a node with key = val
def insert(t, val):
    # [<left>, <right>, <key>, <count>]
    nd = [None, None, val, 1]
    st = []; dr = []
    x = t; y = None
    while x:
        y = x; d = (x[2] <= val)
        st.append(y); dr.append(d)
        x[3] += 1
        x = x[d]

    if not y:
        return nd
    y[d] = nd
    return __splay(st, dr, nd)

# delete a node with key = val
def delete(t, val):
    st = []; dr = []
    x = t
    while x and x[2] != val:
        d = (x[2] <= val)
        st.append(x); dr.append(d)
        x = x[d]
    if not x:
        return t
    t = __splay(st, dr, x)
    return merge(t[0], t[1])

# find a node with key = val
def find(t, val):
    st = []; dr = []
    x = t
    while x and x[2] != val:
        d = (x[2] <= val)
        st.append(x); dr.append(d)
        x = x[d]
    if not x:
        return t, False
    t = __splay(st, dr, x)
    return t, True

# find k-th node in a tree t
def findk(t, k):
    if not t or not 0 < k <= t[3]:
        return t
    x = t
    st = []; dr = []
    while x:
        l = x[0]
        c = (l[3] if l else 0) + 1
 
        if c == k:
            break
 
        st.append(x)
        if c < k:
            k -= c
            x = x[1]
            dr.append(1)
        else:
            x = x[0]
            dr.append(0)
    return __splay(st, dr, x)

# merge a tree l with a tree r
def merge(l, r):
    if not l or not r:
        return l or r
    if not l[1]:
        l[3] += r[3]
        l[1] = r
        return l

    st = []
    x = l
    while x[1]:
        st.append(x)
        x = x[1]

    l = __splay(st, [1]*len(st), x)
    l[3] += r[3]
    l[1] = r
    # update(l)
    return l

# split a tree t into two trees of size k and |t|-k
def split(t, k):
    if not t:
        return None, None
    if not 0 < k < t[3]:
        if k == t[3]:
            return findk(t, k), None
        return None, t
    x = t
    st = []; dr = []
    while x:
        l = x[0]
        c = (l[3] if l else 0) + 1

        if c == k:
            break

        st.append(x)
        if c < k:
            k -= c
            x = x[1]
            dr.append(1)
        else:
            x = x[0]
            dr.append(0)
    l = __splay(st, dr, x)
    r = l[1]
    if r:
        l[3] -= r[3]
    l[1] = None
    # update(l)
    return l, r
